osutil/syncdir.go

```diff
@@ -29,9 +29,29 @@ import (
 	"sort"
 )
 
+type FileStateType string
+
+const (
+	RegularFileStateType FileStateType = "regular"
+	SymlinkFileStateType FileStateType = "symlink"
+)
+
 // FileState is an interface for conveying the desired state of a some file.
 type FileState interface {
 	State() (reader io.ReadCloser, size int64, mode os.FileMode, err error)
+	Type() FileStateType
```

Comment:
os.FileMode can also carry a type, see ModeSymlink, so not sure if we strictly needs this?

Reply:
Updated! Thanks for pointing this out.

Reply:
I updated to returning `mode.Perms()` in State functions to ignore modes passed by mistake and only consider permissions.